File: frontend/lib/models/transaction.dart
import 'dart:convert';
import 'package:json_annotation/json_annotation.dart';

part 'transaction.g.dart';

@JsonSerializable()
class Transaction {
  final int id;
  final int userId;
  final DateTime date;
  final double amount;
  final String description;
  final String? merchant;
  final bool isExpense;
  final bool isRecurring;
  final int? categoryId;
  final String? externalId;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final Category? category;
  final List<Tag> tags;

  Transaction({
    required this.id,
    required this.userId,
    required this.date,
    required this.amount,
    required this.description,
    this.merchant,
    required this.isExpense,
    required this.isRecurring,
    this.categoryId,
    this.externalId,
    required this.createdAt,
    this.updatedAt,
    this.category,
    this.tags = const [],
  });

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
  
  Map<String, dynamic> toJson() => _$TransactionToJson(this);
}

@JsonSerializable()
class Category {
  final int id;
  final String name;
  final String? description;
  final String? color;
  final String? icon;
  final int? userId;
  final DateTime createdAt;

  Category({
    required this.id,
    required this.name,
    this.description,
    this.color,
    this.icon,
    this.userId,
    required this.createdAt,
  });

  factory Category.fromJson(Map<String, dynamic> json) => _$CategoryFromJson(json);
  
  Map<String, dynamic> toJson() => _$CategoryToJson(this);
}

@JsonSerializable()
class Tag {
  final int id;
  final String name;
  final String? description;
  final String? color;
  final int? userId;
  final DateTime createdAt;

  Tag({
    required this.id,
    required this.name,
    this.description,
    this.color,
    this.userId,
    required this.createdAt,
  });

  factory Tag.fromJson(Map<String, dynamic> json) => _$TagFromJson(json);
  
  Map<String, dynamic> toJson() => _$TagToJson(this);
}

@JsonSerializable()
class TransactionCreate {
  final DateTime date;
  final double amount;
  final String description;
  final String? merchant;
  final bool isExpense;
  final bool isRecurring;
  final int? categoryId;
  final String? externalId;
  final List<int>? tagIds;

  TransactionCreate({
    required this.date,
    required this.amount,
    required this.description,
    this.merchant,
    required this.isExpense,
    this.isRecurring = false,
    this.categoryId,
    this.externalId,
    this.tagIds,
  });

  factory TransactionCreate.fromJson(Map<String, dynamic> json) => _$TransactionCreateFromJson(json);
  
  Map<String, dynamic> toJson() => _$TransactionCreateToJson(this);
}

@JsonSerializable()
class TransactionUpdate {
  final DateTime? date;
  final double? amount;
  final String? description;
  final String? merchant;
  final bool? isExpense;
  final bool? isRecurring;
  final int? categoryId;
  final List<int>? tagIds;

  TransactionUpdate({
    this.date,
    this.amount,
    this.description,
    this.merchant,
    this.isExpense,
    this.isRecurring,
    this.categoryId,
    this.tagIds,
  });

  factory TransactionUpdate.fromJson(Map<String, dynamic> json) => _$TransactionUpdateFromJson(json);
  
  Map<String, dynamic> toJson() => _$TransactionUpdateToJson(this);
} 

File: backend/app/schemas/transaction.py
from pydantic import BaseModel, Field
from typing import Optional, List, Union, Dict, Any
from datetime import datetime


class TagBase(BaseModel):
    name: str
    description: Optional[str] = None
    color: Optional[str] = None


class TagCreate(TagBase):
    pass


class TagUpdate(TagBase):
    name: Optional[str] = None


class Tag(TagBase):
    id: str
    user_id: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class CategoryBase(BaseModel):
    name: str
    description: Optional[str] = None
    color: Optional[str] = None
    icon: Optional[str] = None
    type: Optional[str] = "expense"
    parent_id: Optional[str] = None  # ID de la catégorie parente pour les sous-catégories


class CategoryCreate(CategoryBase):
    pass


class CategoryUpdate(CategoryBase):
    name: Optional[str] = None


class Category(CategoryBase):
    id: str
    user_id: Optional[str] = None
    type: Optional[str] = "expense"
    parent_id: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class TransactionBase(BaseModel):
    date: datetime
    amount: float
    description: str
    merchant: Optional[str] = None
    is_expense: Optional[bool] = None  # Peut être None, calculé depuis type si absent
    type: Optional[str] = None  # "income" ou "expense" pour les transactions bancaires
    is_recurring: bool = False
    category_id: Optional[str] = None
    external_id: Optional[str] = None
    explanation: Optional[str] = None


class TransactionCreate(TransactionBase):
    tag_ids: Optional[List[str]] = None


class TransactionUpdate(BaseModel):
    date: Optional[datetime] = None
    amount: Optional[float] = None
    description: Optional[str] = None
    merchant: Optional[str] = None
    is_expense: Optional[bool] = None
    is_recurring: Optional[bool] = None
    category_id: Optional[str] = None
    tag_ids: Optional[List[str]] = None


class Transaction(TransactionBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    category: Optional[Category] = None
    tags: List[Tag] = []
    bank_connection_id: Optional[str] = None  # ID de la connexion bancaire
    bank_account_id: Optional[str] = None  # ID du compte bancaire
    bank: Optional[Dict[str, Any]] = None  # Infos de la banque (ajoutées dynamiquement)
    account: Optional[Dict[str, Any]] = None  # Infos du compte bancaire (ajoutées dynamiquement)

    class Config:
        from_attributes = True


class TransactionWithCategory(Transaction):
    category: Optional[Category] = None


class CategoryAmount(BaseModel):
    category_id: str
    category_name: str
    amount: float
    count: int


class MonthlyReport(BaseModel):
    year: int
    month: int
    total_income: float
    total_expenses: float
    net: float
    expenses_by_category: List[CategoryAmount]
    income_by_category: List[CategoryAmount]


class PeriodReport(BaseModel):
    start_date: datetime
    end_date: datetime
    total_income: float
    total_expenses: float
    net: float
    expenses_by_category: List[CategoryAmount]
    income_by_category: List[CategoryAmount]
    expenses_by_tag: dict
    income_by_tag: dict 